File: hu.bme.mit.inf.ttmc.constraint/src/hu/bme/mit/inf/ttmc/constraint/expr/NeqExpr.java
package hu.bme.mit.inf.ttmc.constraint.expr;

import hu.bme.mit.inf.ttmc.constraint.type.BoolType;
import hu.bme.mit.inf.ttmc.constraint.type.Type;

public interface NeqExpr extends BinaryExpr<Type, Type, BoolType> {
	
	@Override
	public NeqExpr withOps(final Expr<? extends Type> leftOp, final Expr<? extends Type> rightOp);

	@Override
	public NeqExpr withLeftOp(final Expr<? extends Type> leftOp);

	@Override
	public NeqExpr withRightOp(final Expr<? extends Type> rightOp);
}

File: subprojects/common/core/src/main/kotlin/hu/bme/mit/theta/core/type/abstracttype/NeqExpr.kt
/*
 *  Copyright 2025 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hu.bme.mit.theta.core.type.abstracttype

import hu.bme.mit.theta.core.type.BinaryExpr
import hu.bme.mit.theta.core.type.Expr
import hu.bme.mit.theta.core.type.booltype.BoolExprs.Bool
import hu.bme.mit.theta.core.type.booltype.BoolType
import hu.bme.mit.theta.core.utils.TypeUtils.cast
import kotlinx.serialization.Serializable

/** Abstract base class for not-equal (!=) expressions over equational types. */
@Serializable
abstract class NeqExpr<OpType : Equational<OpType>> : BinaryExpr<OpType, BoolType>() {

  companion object {
    private const val OPERATOR_LABEL = "/="

    @JvmStatic
    fun <OpType : Equational<OpType>> create2(leftOp: Expr<*>, rightOp: Expr<*>): NeqExpr<*> {
      @Suppress("UNCHECKED_CAST") val type = leftOp.type as OpType
      val newLeftOp = cast(leftOp, type)
      val newRightOp = cast(rightOp, type)
      return type.Neq(newLeftOp, newRightOp)
    }
  }

  override val operatorLabel: String
    get() = OPERATOR_LABEL

  override val type: BoolType = Bool()
}
